cmd/ftl/cmd_dev.go

```diff
@@ -0,0 +1,258 @@
+package main
+
+import (
+	"bufio"
+	"context"
+	"io/fs"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/TBD54566975/ftl/backend/common/log"
+	"github.com/TBD54566975/ftl/backend/common/moduleconfig"
+	"github.com/TBD54566975/ftl/protos/xyz/block/ftl/v1/ftlv1connect"
+	"github.com/alecthomas/errors"
+	"github.com/bmatcuk/doublestar/v4"
+)
+
+type moduleFolderInfo struct {
+	NumFiles    int
+	LastModTime time.Time
+}
+
+type devCmd struct {
+	BaseDir string        `arg:"" help:"Directory to watch for FTL modules" type:"existingdir" default:"."`
+	Watch   time.Duration `help:"Watch template directory at this frequency and regenerate on change." default:"500ms"`
+	modules map[string]moduleFolderInfo
+	client  ftlv1connect.ControllerServiceClient
+}
+
+func (d *devCmd) Run(ctx context.Context, client ftlv1connect.ControllerServiceClient) error {
+	logger := log.FromContext(ctx)
+	logger.Infof("Watching %s for FTL modules", d.BaseDir)
+
+	d.modules = make(map[string]moduleFolderInfo)
+	d.client = client
+
+	lastScanTime := time.Now()
+	for {
+		iterationStartTime := time.Now()
+
+		tomls, err := d.getTomls()
+		if err != nil {
+			return errors.WithStack(err)
+		}
+
+		d.addOrRemoveModules(tomls)
+
+		for dir := range d.modules {
+			currentModule := d.modules[dir]
+			err := d.updateFileInfo(dir)
+			if err != nil {
+				return errors.WithStack(err)
+			}
+
+			if currentModule.NumFiles != d.modules[dir].NumFiles || d.modules[dir].LastModTime.After(lastScanTime) {
+				deploy := deployCmd{
+					Replicas:  1,
+					ModuleDir: dir,
+				}
+				err = deploy.Run(ctx, client)
+				if err != nil {
+					logger.Errorf(err, "Error deploying module %s. Will retry", dir)
+					delete(d.modules, dir)
```

Comment:
If a deploy fails, I'm just deleting that module so that it is re detected next time around. This helps with modules that are dependent on other modules. It might take a few loops to get everything running, but it does get going eventually.